### PR TITLE
os/bluestore: Switch to time-based adaptive near-fit alogrithm

### PR DESCRIPTION
**Pull request by markhpc**
Previously the AVL allocator had no limits on how long near-fit mode could run.  This resulting in the AVL allocator doing a very good job searching for contiguous space but when the disk was heavily fragmented could result in extremely long allocation requests as documented in https://tracker.ceph.com/issues/52804 and https://github.com/ceph/ceph/pull/45755.  To remedy this, byte and count limits were added in #41615 which prevented long allocation requests.  Unfortunately these limits quickly resulted in fragmentation that hurt large write performance on the Samsung PM983 drives in our test cluster.  The following tests are only short 30s iterations, but demonstrate the problem:

### Quincy Off (search_count=0, search_bytes=0MB)

Iteration | 4KB (IOPS) | 128KB (MiB/s) | 4MB (MiB/s)
-- | -- | -- | --
0 | 79092 | 1766 | 1812
1 | 73365 | 1784 | 1814
2 | 74780 | 1808 | 1816

mako-quincy-406c2368-alloc_ff0-4MBrandwrite_sm

### Quincy Default (search_count=100, search_bytes=16MB)

Iteration | 4KB (IOPS) | 128KB (MiB/s) | 4MB (MiB/s)
-- | -- | -- | --
0 | 80492 | **1680** | 1818
1 | 77406 | **1711** | **1671**
2 | 75956 | **906.7** | **1350**

mako-quincy-406c2368-4MBrandwrite_sm

Choosing larger values for the count and byte limits helps to some extent, but it's difficult to guess at what those defaults should be, and even an 8x increase over the defaults does not get us back to previous performance levels and may actually hurt performance when set too high:

### Quincy 4X (search_count=400, search_bytes=64MB)

Iteration | 4KB (IOPS) | 128KB (MiB/s) | 4MB (MiB/s)
-- | -- | -- | --
0 | 79769 | 1802 | 1773
1 | **72407** | 1786 | 1802
2 | **72319** | 1792 | **1601**

mako-quincy-406c2368-alloc_ff4x-4MBrandwrite_sm

### Quincy 8X (search_count=800, search_bytes=128MB)

Iteration | 4KB (IOPS) | 128KB (MiB/s) | 4MB (MiB/s)
-- | -- | -- | --
0 | 79488 | 1797 | **1474**
1 | **70332** | 1785 | **1563**
2 | **71895** | 1785 | **1608**

mako-quincy-406c2368-alloc_ff8x-4MBrandwrite_sm

To combat this issue, this PR replaces the count and byte limits from #41615 with a single search_time parameter that indicates the maximum amount of time that the AVL allocator should stay in near-first mode before switching to best-first.  It also has an optimization to reduce the number of times we call ceph_time_now() when quickly searching for allocations.  This single value is easier to tune than the dual count/byte limits and easier to think about in terms of the near-first contribution to total allocator lookup time.

When nf_search_time is lower than the time it takes for the allocator to stay in near-first, allocations will be fragmented (though not as badly) like in the current code:

### Adaptive NF (nf_search_time=500us)

Iteration | 4KB (IOPS) | 128KB (MiB/s) | 4MB (MiB/s)
-- | -- | -- | --
0 | 71975 | 1793 | 1776
0 | **68608** | 1808 | 1799
0 | 76577 | 1787 | **1713**

mako-quincy-406c2368-alloc_aff2_0500us-4MBrandwrite_sm

However setting it higher (1ms) results in a similar allocation pattern to when we disable search_count and search_bytes in the current code:

### Adaptive NF (nf_search_time=1ms)

Iteration | 4KB (IOPS) | 128KB (MiB/s) | 4MB (MiB/s)
-- | -- | -- | --
0 | 79504 | 1801 | 1830
1 | **68696** | 1795 | 1832
2 | 72557 | 1794 | 1829

mako-quincy-406c2368-alloc_aff2_1000us-4MBrandwrite_sm

Setting it even higher (2ms) may provide more headroom to allow for allocator stalls without fragmenting, but oddly may have an initial (but not permanent?) impact on 4k random write performance.  This would likely be the case with the current code if search_count and/or search_bytes were set much higher.

### Adaptive NF (nf_search_time=2ms)

Iteration | 4KB (IOPS) | 128KB (MiB/s) | 4MB (MiB/s)
-- | -- | -- | --
0 | **52731** | 1766 | 1831
1 | **68494** | 1815 | 1816
2 | 77683 | 1798 | 1830

mako-quincy-406c2368-alloc_aff2_2000us-4MBrandwrite_sm

@ifed01 I don't have the the bin file you used in your ceph_test_alloc_replay test.  If you think this PR makes sense, would you want to try it?  Given that I'm limiting near-first search to 1ms, I think it makes sense that it wouldn't prevent what I think is the 746us allocation from your output in #45755?

Signed-off-by: Mark Nelson <mnelson@redhat.com>

### CODE REVIEW

**Comment by ifed01** [comment]
@markhpc - so I double checked the time units in replay output - it's ns indeed. So in the output I shared we're talking about usec, not ms. 
And I gave a short try to your PR in my replay test and it looks like 1ms is a way to high. Although I still need to arrange all the things with these experiments it looks like 1-5 usec is a better option to match stupid allocator numbers...

**Comment by markhpc** [comment]
@ifed01 I initially tried much lower thresholds, but anything much lower than 1ms results in scattered IO (as you can see in the 500us test) and lower throughput.  I'm guessing that once we start doing best-fit it's tough to get back into a fast state under load.  Having said that, it's clear that larger thresholds are showing somewhat better behavior than lower thresholds even when going into best-fit.

I think the issue we need to tackle right now for Quincy is the 75-100ms numbers you previously saw.  We can't allow near-fit to spend that much time doing searches.  Limiting near-fit to ~1ms should take care of the majority of that.  The trade-off below the 1ms level between fast searches and fragmentation is still open, but at least in these tests it looks like once AVL starts using best-fit it's going to write IOs across the entire disk and these Samsung NVMe drives at least seem to hate that.

I'd argue that for the quincy release we apply this PR with a high threshold (~1ms) to deal with the worst case numbers you reported and then we do a deep dive to figure out if it's reasonable that near-fit is taking this long to fulfill large allocation requests.  It may be the Hybrid/AVL as it currently exists isn't the best general solution (though I will say that for small random write IOPs, hybrid appears to be quite fast).

**Comment by markhpc** [comment]
Pushed some changes, including setting the cursor position when in best-fit mode.  Josh and I walked through the allocator code earlier today and noticed that near-fit will set the cursor position, but if we short circuit out (either via the changes in #41615 or in the earlier version of this code) will simply restart the near-fit search at the same cursor position as it searched previously and do this for every single allocation attempt.

Currently this code also has debugging at debug0 that should be removed before this PR merges.

**Comment by markhpc** [comment]
Attaching debug log results for the 4kb/128kb/4mb 3 iteration cycle tests.  Debug log lines are recorded randomly at 1/1000th probability.  Allocations will not show up as contiguous due to gaps, but trends can be seen in the results.
[AvlAllocator-ANF_500us_debug.txt.gz](https://github.com/ceph/ceph/files/8448010/AvlAllocator-ANF_500us_debug.txt.gz)

**Comment by ifed01** [comment]
> Pushed some changes, including setting the cursor position when in best-fit mode. Josh and I walked through the allocator code earlier today and noticed that near-fit will set the cursor position, but if we short circuit out (either via the changes in #41615 or in the earlier version of this PR) it will simply restart the near-fit search at the same cursor position as it searched previously and do this for every single allocation attempt never making any progress.
> 
> Currently this code also has debugging at debug0 that should be removed before this PR merges.

That's a good finding about cursor position. Just a comment/proposal - IMO better approach would be to increment cursor position on any lookup in near-fit mode, both successful and non-successful. This will enforce incremental lookup over the whole space. While setting cursor using lookup result from best-fit makes a little sense since it performs scanning primarily using chunk sizes and hence returned cursor position is [semi-]random.

**Comment by markhpc** [comment]
> That's a good finding about cursor position. Just a comment/proposal - IMO better approach would be to increment cursor position on any lookup in near-fit mode, both successful and non-successful. This will enforce incremental lookup over the whole space. While setting cursor using lookup result from best-fit makes a little sense since it performs scanning primarily using chunk sizes and hence returned cursor position is [semi-]random.

It turns out that it seems to depend on the length of time we spend in near-fit mode.  For very short periods it may be better to use best-fit as the starting point.  But the near-fit change is literally a 1 line of code change so we might as well just use that.  I think for quincy we can merge that and it might be enough to get us back into good behavior.  I think we probably still want this PR since it makes tuning more straightforward, but we can decide that later.  Anyway, results from testing are here:

https://docs.google.com/spreadsheets/d/18jFchgrl1tSpQsv0rwPnKsilbBaBl0DZFmgkIG5nwrE/edit#gid=805410009

File: src/os/bluestore/AvlAllocator.cc [hunk 1]
```diff
@@ -57,22 +85,48 @@ uint64_t AvlAllocator::_pick_block_after(uint64_t *cursor,
...
   // If we reached end, start from beginning till cursor.
   for (auto rs = range_tree.begin(); rs != rs_start; ++rs) {
+    auto sb2 = rs->start - rs_start->start;
+
     uint64_t offset = p2roundup(rs->start, align);
     if (offset + size <= rs->end) {
-      *cursor = offset + size;
+      auto r = ((double) rand() / (RAND_MAX));
+      if (r < 0.001) {
+        utime_t cur_time = ceph_clock_now();
```
[the remark this answers is not in the document]
Comment (markhpc) [reply src/os/bluestore/AvlAllocator.cc:94]: @ifed01 Yep, I agree!  If only for the guarantee that it always moves forward.  Lots of discussion from 2016 here:
https://lore.kernel.org/all/56A4BB20.1040305@digiware.nl/T/

I also spent some time looking for clock overhead benchmarking.  There is an old blog post from a self-confessed wallstreet programmer that wrote a fairly reasonable looking (to me at least) test program here (taskset cpu pinning, clean code, etc):

https://btorpey.github.io/blog/2014/02/18/clock-sources-in-linux/

Quoting from his post (results in ns):

```
ClockBench.cpp
                   Method       samples     min     max     avg  median   stdev
           CLOCK_REALTIME       255       54.00   58.00   55.65   56.00    1.55
    CLOCK_REALTIME_COARSE       255        0.00    0.00    0.00    0.00    0.00
          CLOCK_MONOTONIC       255       54.00   58.00   56.20   56.00    1.46
      CLOCK_MONOTONIC_RAW       255      650.00 1029.00  690.35  839.50   47.34
   CLOCK_MONOTONIC_COARSE       255        0.00    0.00    0.00    0.00    0.00
              cpuid+rdtsc       255       93.00   94.00   93.23   93.50    0.42
                    rdtsc       255       24.00   28.00   25.19   26.00    1.50
Using CPU frequency = 2.660000
```
Code is available here: https://github.com/btorpey/clocks

Running this on mako yields:
```
ClockBench.cpp
                   Method	samples	    min	    max	    avg	 median	  stdev
           CLOCK_REALTIME	1023	  20.00	  31.00	  25.35	  25.50	   5.02
    CLOCK_REALTIME_COARSE	1023	   0.00	   0.00	   0.00	   0.00	   0.00
          CLOCK_MONOTONIC	1023	  20.00	  31.00	  25.34	  25.50	   5.02
      CLOCK_MONOTONIC_RAW	1023	  20.00	  31.00	  25.34	  25.50	   5.02
   CLOCK_MONOTONIC_COARSE	1023	   0.00	   0.00	   0.00	   0.00	   0.00
              cpuid+rdtsc	1023	  39.00	  51.00	  42.14	  45.00	   4.11
                   rdtscp	1023	  19.00	  31.00	  20.63	  25.00	   2.44
                    rdtsc	1023	   8.00	  21.00	  11.06	  14.50	   3.11
Using CPU frequency = 2.245665

```

And on officinalis:
```
ClockBench.cpp
                   Method	samples	    min	    max	    avg	 median	  stdev
           CLOCK_REALTIME	1023	  15.00	  17.00	  16.00	  16.00	   0.51
    CLOCK_REALTIME_COARSE	1023	   0.00	   0.00	   0.00	   0.00	   0.00
          CLOCK_MONOTONIC	1023	  15.00	  17.00	  16.00	  16.00	   0.45
      CLOCK_MONOTONIC_RAW	1023	  16.00	  18.00	  16.50	  17.00	   0.59
   CLOCK_MONOTONIC_COARSE	1023	   0.00	   0.00	   0.00	   0.00	   0.00
              cpuid+rdtsc	1023	  24.00	  32.00	  28.75	  28.00	   2.22
                   rdtscp	1023	   4.00	  12.00	   9.80	   8.00	   2.36
                    rdtsc	1023	   0.00	   8.00	   6.60	   4.00	   2.01
Using CPU frequency = 2.200000

```

Micro-benchmarking is notoriously tricky but this lines up pretty well with what I'm seeing in actual testing (basically zero overhead from clock lookup with the current backoff scheme).  eyeballing the numbers I saw earlier I think we're averaging ~200ns+ per iteration with the current code, but the vast majority of that seems to be tied up in walking the AVL tree.  Based on these results it's looking like we could save 15-26ns in the fast case if we pulled the first lookup out of the loop before we did the initial time calculation.  I can do that fairly easily if folks think it's worth the extra code complexity.

In either case, figure that for a 10us limit we are typically iterating maybe 56-120 times with 4-5 time lookups.  (initial + 8 + 16 + 32 + maybe 64) with something like 60-125ns aggregate lookup time for a rough 10-15us overall near-fit search limit.  IE that's about 1% or less of the lookup time is in clock checks.  We could probably play some games to lower this but I don't think this is going to be a major source of overhead for some time.

[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/bluestore/AvlAllocator.cc:94]: @markhpc - thanks for the super thorough response! Just run the same benchmark in my lab - pretty similar results. So actually there are no performance difference between clock implementations we use.

File: src/os/bluestore/AvlAllocator.cc [hunk 2]
```diff
@@ -342,10 +395,13 @@ AvlAllocator::AvlAllocator(CephContext* cct,
     cct->_conf.get_val<uint64_t>("bluestore_avl_alloc_bf_threshold")),
   range_size_alloc_free_pct(
     cct->_conf.get_val<uint64_t>("bluestore_avl_alloc_bf_free_pct")),
-  max_search_count(
-    cct->_conf.get_val<uint64_t>("bluestore_avl_alloc_ff_max_search_count")),
-  max_search_bytes(
-    cct->_conf.get_val<Option::size_t>("bluestore_avl_alloc_ff_max_search_bytes")),
+  nf_search_time(
```
Comment (ifed01) [inline src/os/bluestore/AvlAllocator.cc:398]: If we stick with the legacy config parameter access (ie cct->_conf->bluestore_avl_alloc_nf_search_time) we can track their modifications with a zero overhead in real-time rather then exclusively on startup. Hence one would be able to tune them without OSD restart...

**Comment by ifed01** [comment]
jenkins test make check

File: src/os/bluestore/AvlAllocator.cc [hunk 3]
```diff
@@ -82,6 +83,7 @@ uint64_t AvlAllocator::_pick_block_fits(uint64_t size,
   for (auto rs = rs_start; rs != range_size_tree.end(); ++rs) {
     uint64_t offset = p2roundup(rs->start, align);
     if (offset + size <= rs->end) {
+      *cursor = offset + size;
```
[the remark this answers is not in the document]
Comment (markhpc) [reply src/os/bluestore/AvlAllocator.cc:139]: interestingly there is some trade-off where always updating the cursor in near-fit is better with low search time and updating the cursor in best-fit seems to be better for high search time.  I think we can just do it always in near-fit though.  The big win is just avoiding the repeated searches.

File: src/os/bluestore/AvlAllocator.cc [hunk 4]
```diff
@@ -35,20 +35,48 @@ uint64_t AvlAllocator::_pick_block_after(uint64_t *cursor,
...
-    if (search_bytes = rs->start - rs_start->start;
-	max_search_bytes > 0 && search_bytes > max_search_bytes) {
-      return -1ULL;
+    // If we are taking too long, give up and switch to best-fit
+    if (nf_search_time > 0 && ++search_count > max_search) {
+      utime_t cur_time = ceph_clock_now();
+      auto search_time = cur_time - start_time;
+      if (search_time > nf_search_time) {
+        nf_misses++;
+        ldout(cct, 0) << __func__ << " search_time: " << search_time
```
[the remark this answers is not in the document]
Comment (markhpc) [reply src/os/bluestore/AvlAllocator.cc:68]: yeah, throwaway when we remove DNM, see comment when tag was added

File: src/os/bluestore/AvlAllocator.cc (hunk 2 above)
Comment (markhpc) [reply src/os/bluestore/AvlAllocator.cc:398]: yeah, that's a good point.  I think we can do it with the new style with some trickery right?

Comment (markhpc) [reply src/os/bluestore/AvlAllocator.cc:398]: FWIW I *like* using the legacy interface and hate how fragmented it all has become, but I feel a little guilty for keeping using it. :)  Having said that, I'm not at all opposed to using it especially in the hot path if other people feel the same way.

**Pull request closed without merging**